src/com/amazon/kinesis/streaming/agent/providers/CLIWebIdentityTokenCredentialsProvider.java

```diff
@@ -0,0 +1,108 @@
+package com.amazon.kinesis.streaming.agent.providers;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.DateTimeFormatterBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.amazonaws.auth.AWSCredentials;
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.auth.BasicSessionCredentials;
+
+/**
+ * This provider uses the AWS CLI to call assume-role-with-web-identity to retrieve credentials.
+ * Refresh on expiration is implemented.
+ */
+public class CLIWebIdentityTokenCredentialsProvider implements AWSCredentialsProvider {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CLIWebIdentityTokenCredentialsProvider.class);
+    private static final DateTimeFormatter DATE_TIME_FORMATTER = new DateTimeFormatterBuilder()
+        .appendPattern("yyyy-MM-dd'T'HH:mm:ss")
+        .appendTimeZoneOffset("+00:00", true, 2, 4)
+        .toFormatter();
+
+    private final String roleArn = System.getenv("AWS_ROLE_ARN");
+    private String webIdentityToken = "";
+
+    private String accessKey = "";
+    private String secretKey = "";
+    private String sessionToken = "";
+    private DateTime expiration = DateTime.now();
+
+    public CLIWebIdentityTokenCredentialsProvider() {
+        this.refresh();
+        if (accessKey.isEmpty() || secretKey.isEmpty() || sessionToken.isEmpty()) {
+            LOGGER.warn("Error instantiating CLI Web Identity Token credentials provider. Falling back to the standard provider.");
+        }
+    }
+
+    @Override
+    public AWSCredentials getCredentials() {
+        if (accessKey.isEmpty() || secretKey.isEmpty() || sessionToken.isEmpty()) {
+            throw new IllegalStateException("Unable to retrieve Web Token Identity credentials from CLI.");
+        }
+
+        if (DateTime.now(DateTimeZone.UTC).getMillis() >= expiration.getMillis()) {
+            this.refresh();
+        }
+
+        return new BasicSessionCredentials(this.accessKey, this.secretKey, this.sessionToken);
+    }
+
+    @Override
+    public void refresh() {
+        try {
+            LOGGER.info("Refreshing Web Identity Token credentials from CLI...");
+            this.getToken();
+            final String cmd = "aws sts assume-role-with-web-identity --role-session-name kinesis_agent --role-arn "
+                + this.roleArn + " --web-identity-token " + this.webIdentityToken + " --query Credentials --output yaml";
+            this.getAssumedCredentials(Runtime.getRuntime().exec(cmd));
+        } catch (IOException e) {
```

Comment:
do you need to handle other exceptions? like this cmd is assuming you already have an aws creds config what's able to perform `sts assume-role`, what if that creds itself is missing?

Reply:
I've tested it both with CLI missing and token file missing, IOException is the only thing thrown. This makes sense the process output is read as a stream.